File: LeetCode/困难/LCR 170. 交易逆序对的总数/SolutionLCR170.java
/*
url: https://leetcode.cn/problems/shu-zu-zhong-de-ni-xu-dui-lcof/description/
LeetCode参考: https://leetcode.cn/problems/shu-zu-zhong-de-ni-xu-dui-lcof/solutions/216984/shu-zu-zhong-de-ni-xu-dui-by-leetcode-solution/
相关: LeetCode493. 翻转对
标签: 逆序对数, 归并排序, 树状数组
*/

public class SolutionLCR170 {
    int ans = 0;
    int[] temp;

    public int reversePairs(int[] record) {
        if (record.length == 0) return 0;
        this.temp = new int[record.length / 2]; // 用作merge的数组
        mergeSort(record, 0, record.length);
        return ans;
    }

    private void mergeSort(int[] record, int left, int right) {
        if (right - left == 1) return;
        int mid = (left + right) / 2;
        mergeSort(record, left, mid);
        mergeSort(record, mid, right);
        merge(record, left, mid, right);
    }

    private void merge(int[] record, int left, int mid, int right) {
        for (int i = left; i < mid; i++) {
            temp[i - left] = record[i];
        }
        int p1 = 0, p2 = mid, p = left;
        // (mid - left)为前半段长度
        while (p1 < (mid - left) && p2 < right) {
            if (temp[p1] <= record[p2]) {
                record[p++] = temp[p1++];
            }
            else {
                ans += (mid - left) - p1; // 此时temp[p1]及右边的元素全部大于record[p2]，构成逆序对
                record[p++] = record[p2++];
            }
        }
        while (p1 < (mid - left)) {
            record[p++] = temp[p1++];
        }
    }

    public static void main(String[] args) {
        SolutionLCR170 solu = new SolutionLCR170();
        System.out.println(solu.reversePairs(new int[]{9, 7, 5, 4, 6}));
    }
}

File: LeetCode/困难/LCR 170. 交易逆序对的总数/solutionLCR170.cpp
#include <bits/stdc++.h>

using namespace std;

class SolutionLCR170 {
public:
    int ans = 0;
    vector<int> temp; // 用作归并的数组

    /* 把线性的结构树化，用归并排序或者认为是线段树的分治思路。
    若对A B两个子向量归并的一步选A，则没有逆序对，若选B，则B的这个头能与|A|都形成逆序对，直接计数，不用反复比较
    */
    int reversePairs(vector<int>& nums) {
        int n = nums.size();
        temp.resize(n / 2);
        merge_sort(nums, 0, n);
        return ans;
    }

    void merge_sort(vector<int>& nums, int left, int right) {
        if (right - left <= 1) return;
        int mid = (left + right) >> 1;
        merge_sort(nums, left, mid);
        merge_sort(nums, mid, right);
        merge(nums, left, mid, right);
    }

    void merge(vector<int>& nums, int left, int mid, int right) {
        for (int i = left; i < mid; i++) temp[i - left] = nums[i];

        int i = 0, j = mid, k = left;
        while (i < mid - left && j < right) {
            // 选左边
            if (temp[i] <= nums[j]) {
                nums[k++] = temp[i++];
            }
            // 选右边
            else {
                ans += mid - left - i;
                nums[k++] = nums[j++];
            }
        }
        // 处理剩余的左部。剩余的右部不用处理
        while (i < mid - left) {
            nums[k++] = temp[i++];
        }
    }
};

int main(int argc, char const *argv[]) {
    SolutionLCR170 solu;
    cout << solu.reversePairs(*new vector<int>{9, 7, 5, 4, 6}) << endl;
    return 0;
}